File: M/729-my-calendar-i/main.go
package main

import "fmt"

type MyCalendar struct {
	Start []int
	End   []int
}

func Constructor() MyCalendar {
	return MyCalendar{Start: []int{}, End: []int{}}
}

func (this *MyCalendar) Book(start int, end int) bool {
	if len(this.Start) == 0 {
		this.Start = append(this.Start, start)
		this.End = append(this.End, end)
		return true
	}
	idx := -1
	for i := 0; i < len(this.Start); i++ {
		if this.Start[i] > start {
			idx = i
			break
		}
	}
	if idx == -1 {
		if this.End[len(this.End)-1] <= start {
			this.Start = append(this.Start, start)
			this.End = append(this.End, end)
			return true
		} else {
			return false
		}
	} else {
		if idx == 0 {
			if this.Start[0] >= end {
				this.Start = append([]int{start}, this.Start...)
				this.End = append([]int{end}, this.End...)
				return true
			} else {
				return false
			}
		} else {
			if this.End[idx-1] <= start && this.Start[idx] >= end {
				s, e := make([]int, len(this.Start)+1), make([]int, len(this.End)+1)
				copy(s, this.Start[0:idx])
				copy(e, this.End[0:idx])
				s[idx] = start
				e[idx] = end
				copy(s[idx+1:], this.Start[idx:])
				copy(e[idx+1:], this.End[idx:])

				this.Start = s
				this.End = e
				return true
			} else {
				return false
			}
		}
	}
}

func main() {
	fmt.Println("a")
}

File: M/729-my-calendar-i/main.rs
struct Solution {}

use std::collections::HashMap;
struct MyCalendar {
  T: HashMap<i32, bool>,
  L: HashMap<i32, bool>,
}

impl MyCalendar {
  fn new() -> Self {
    MyCalendar {
      T: HashMap::new(),
      L: HashMap::new(),
    }
  }

  fn query(&self, start: i32, end: i32, l: i32, r: i32, idx: i32) -> bool {
    if r < start || end < l {
      return false;
    }

    if self.L.contains_key(&idx) && *self.L.get(&idx).unwrap() {
      return true;
    }

    if start <= l && r <= end {
      return self.T.contains_key(&idx) && *self.T.get(&idx).unwrap();
    }

    let mid = (l + r) >> 1;
    self.query(start, end, l, mid, idx * 2) || self.query(start, end, mid + 1, r, idx * 2 + 1)
  }
  fn update(&mut self, start: i32, end: i32, l: i32, r: i32, idx: i32) {
    if r < start || end < l {
      return;
    }
    if start <= l && r <= end {
      self.T.insert(idx, true);
      self.L.insert(idx, true);
    } else {
      let mid = (l + r) >> 1;
      self.update(start, end, l, mid, idx * 2);
      self.update(start, end, mid + 1, r, idx * 2 + 1);
      self.T.insert(idx, true);
      if self.L.contains_key(&(idx * 2))
        && *self.L.get(&(idx * 2)).unwrap()
        && self.L.contains_key(&(idx * 2 + 1))
        && *self.L.get(&(idx * 2 + 1)).unwrap()
      {
        self.L.insert(idx, true);
      }
    }
  }

  fn book(&mut self, start_time: i32, end_time: i32) -> bool {
    let start = start_time;
    let end = end_time;
    if self.query(start, end - 1, 0, 1000000000, 1) {
      return false;
    }
    self.update(start, end - 1, 0, 1000000000, 1);
    true
  }
}
